### PR TITLE
Modded ItemInstance Data and TileInteractionCallback

### PR DESCRIPTION
**Pull request by valoeghese**
~~I just stole the mixin from my teleporters mod but made it fancy. Haven't tested the api but it probably works idk.~~
I tested and fixed some stuff

### CODE REVIEW

File: src/main/java/io/github/minecraftcursedlegacy/impl/data/DataStorage.java [hunk 1]
```diff
@@ -0,0 +1,13 @@
+package io.github.minecraftcursedlegacy.impl.data;
+
+import java.util.function.Supplier;
+
+import io.github.minecraftcursedlegacy.api.data.ModdedData;
+import io.github.minecraftcursedlegacy.api.registry.Id;
+import net.minecraft.util.io.CompoundTag;
+
+public interface DataStorage {
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/impl/data/DataStorage.java:11]: why's this `Data` but everything else `ModdedData`? (honestly I prefer `Data` over `ModdedData`)

**Comment by valoeghese** [comment]
Todo:
- mixin Id21Packet, ClientPlayNetworkHandler#method_1453? (Client side stuff it seems)
- other networking stuff
- why do item entities stuff up

File: src/main/java/io/github/minecraftcursedlegacy/api/data/AttachedData.java [hunk 2]
```diff
@@ -0,0 +1,27 @@
+package io.github.minecraftcursedlegacy.api.data;
+
+import io.github.minecraftcursedlegacy.api.registry.Id;
+import net.minecraft.util.io.CompoundTag;
+
+/**
+ * Data which can be attached to various vanilla objects, such as items and blocks.
+ * @see {@link DataManager}.
+ */
+public interface AttachedData {
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/api/data/AttachedData.java:10]: attached data good and best in game

File: src/main/java/io/github/minecraftcursedlegacy/api/data/DataManager.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+import net.minecraft.util.io.CompoundTag;
+
+/**
+ * Manager for data which can be attached to various vanilla objects, such as items and blocks.
+ */
+public final class DataManager<T> {
+	private DataManager() {
+	}
+
+	private final Map<Id, Function<T, ? extends AttachedData>> moddedDataFactories = new HashMap<>();
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/api/data/DataManager.java]: `moddedDataFactories` -> `attachedDataFactories`

File: src/main/java/io/github/minecraftcursedlegacy/api/event/TileInteractionCallback.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+import net.minecraft.entity.player.Player;
+import net.minecraft.item.ItemInstance;
+import net.minecraft.level.Level;
+import net.minecraft.tile.Tile;
+
+/**
+ * Callback for right clicking a tile. This is run both client and server side.
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/api/event/TileInteractionCallback.java:13]: Nice javadoc

File: src/main/java/io/github/minecraftcursedlegacy/api/data/DataManager.java [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+	/**
+	 * Adds the specified attached data to the {@link DataManager} instance. This data can later be accessed on an instance of the object via {@link #getAttachedData}.
+	 * @return a key to use to retrieve the attached data from an object.
+	 */
+	public <E extends AttachedData> DataKey<E> addModdedData(Id id, Function<T, E> dataProvider) {
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/api/data/DataManager.java]: `addModdedData` -> `addAttachedData`

File: src/main/java/io/github/minecraftcursedlegacy/mixin/MixinItemInstance.java [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+import io.github.minecraftcursedlegacy.api.data.DataManager;
+import io.github.minecraftcursedlegacy.api.registry.Id;
+import io.github.minecraftcursedlegacy.impl.data.DataStorage;
+import net.minecraft.item.ItemInstance;
+import net.minecraft.util.io.CompoundTag;
+
+@Mixin(ItemInstance.class)
+public abstract class MixinItemInstance implements DataStorage {
+	private CompoundTag api_data = new CompoundTag();
+	private Map<Id, AttachedData> api_moddedDataMap = new HashMap<>();
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/mixin/MixinItemInstance.java]: `api_moddedDataMap` -> `api_attachedDataMap`

File: src/main/java/io/github/minecraftcursedlegacy/impl/data/DataStorage.java [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+import java.util.function.Supplier;
+
+import io.github.minecraftcursedlegacy.api.data.AttachedData;
+import io.github.minecraftcursedlegacy.api.registry.Id;
+import net.minecraft.util.io.CompoundTag;
+
+public interface DataStorage {
+	CompoundTag getModdedTag();
+	void putModdedData(Id id, AttachedData data);
+	AttachedData getModdedData(Id id, Supplier<AttachedData> supplier);
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/impl/data/DataStorage.java]: `getModdedData` and `putModdedData` -> `getAttachedData` and `putAttachedData`

Comment (valoeghese) [reply src/main/java/io/github/minecraftcursedlegacy/impl/data/DataStorage.java]: ah, nitpicking the impl names as well as the api.

Comment (williambl) [reply src/main/java/io/github/minecraftcursedlegacy/impl/data/DataStorage.java]: nothing is safe from the bikeshed.

**Comment by valoeghese** [comment]
Found It: I'm gonna inject to the INVOKE_ASSIGN (I think that's the right inject place) of `PlayerInventory.method_685`, since I need to mess with the object that's constructed by appending the attached data. Then that'll be it I think.

**Comment by valoeghese** [comment]
Heck I closed it

**Comment by valoeghese** [comment]
except for packets, but who cares about multiplayer lol

**Review by williambl — CHANGES_REQUESTED** [review]
I care about multiplayer

**Review by Ueaj-Kerman — CHANGES_REQUESTED** [review]
h

**Review by CoolMineman — CHANGES_REQUESTED** [review]
make work in multiplayer k thx

**Comment by valoeghese** [comment]
I might merge this and do the packet sync later...
there are about 7 packets that transfer item instances or sth. WillBL wants me to make a new packet to send alongside, however I'm not sure how that would work.

**Comment by williambl** [comment]
Okay on second thoughts I'm not really sure we need this to sync. If someone really needs it to sync they can do it themselves. All the actual logic should be done serverside imo.

**Comment by valoeghese** [comment]
Can we have final code reviews, then?

File: src/main/java/io/github/minecraftcursedlegacy/api/event/TileInteractionCallback.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+				return ActionResult.PASS;
+			});
+
+	/**
+	 * @param player the player causing the tile interaction.
+	 * @param level the level the tile is being interacted with in.
+	 * @param item the item instance that the player is using to interact with the tile.
+	 * @param tile the tile being interacted with at the time of this event firing. This does not change if an event subscriber alters the tile at that position.
+	 * @param i1 I have no idea what this is but its the same as the last parameter of {@link ItemInstance#useOnTile(Player, Level, int, int, int, int)};
```
Comment (Juuxel) [inline src/main/java/io/github/minecraftcursedlegacy/api/event/TileInteractionCallback.java]: Probably the block face?

File: src/main/java/io/github/minecraftcursedlegacy/mixin/MixinItemEntity.java [hunk 9]
```diff
@@ -0,0 +1,9 @@
+package io.github.minecraftcursedlegacy.mixin;
+
+import org.spongepowered.asm.mixin.Mixin;
+
+import net.minecraft.entity.ItemEntity;
+
+@Mixin(ItemEntity.class)
+public class MixinItemEntity {
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/mixin/MixinItemEntity.java]: Can we get rid of this?

File: src/main/java/io/github/minecraftcursedlegacy/mixin/MixinClientInteractionManager.java [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+import io.github.minecraftcursedlegacy.impl.event.TileInteractionImpl;
+import net.minecraft.client.ClientInteractionManager;
+import net.minecraft.entity.player.Player;
+import net.minecraft.item.ItemInstance;
+import net.minecraft.level.Level;
+
+@Mixin(ClientInteractionManager.class)
+public class MixinClientInteractionManager {
+	@Inject(at = @At("HEAD"), method = "useItemOnTile", cancellable = true)
+	private void onTileInteract(Player player, Level level, ItemInstance item, int x, int y, int z, int i1, CallbackInfoReturnable<Boolean> info) {
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/mixin/MixinClientInteractionManager.java]: api_ prefix?

File: src/main/java/io/github/minecraftcursedlegacy/mixin/MixinClass_70.java [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+import io.github.minecraftcursedlegacy.impl.event.TileInteractionImpl;
+import net.minecraft.entity.player.Player;
+import net.minecraft.item.ItemInstance;
+import net.minecraft.level.Level;
+import net.minecraft.server.ServerInteractionManager;
+
+@Mixin(ServerInteractionManager.class)
+public class MixinClass_70 {
+	@Inject(at = @At("HEAD"), method = "activateTile", cancellable = true)
+	private void onTileInteract(Player player, Level level, ItemInstance item, int x, int y, int z, int i1, CallbackInfoReturnable<Boolean> info) {
```
Comment (williambl) [inline src/main/java/io/github/minecraftcursedlegacy/mixin/MixinClass_70.java]: api_ prefix?

File: src/test/java/io/github/minecraftcursedlegacy/testmixin/MixinPlayer.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package io.github.minecraftcursedlegacy.testmixin;
+
+import org.spongepowered.asm.mixin.Mixin;
+
+import net.minecraft.entity.player.Player;
+
+@Mixin(Player.class)
+public class MixinPlayer {
```
Comment (williambl) [inline src/test/java/io/github/minecraftcursedlegacy/testmixin/MixinPlayer.java:8]: Can we get rid of this?

**Review by williambl — CHANGES_REQUESTED** [review]
Just a couple of things. Otherwise looks ready to merge :)

File: src/test/java/io/github/minecraftcursedlegacy/testmixin/MixinPlayer.java (hunk 12 above)
Comment (valoeghese) [reply src/test/java/io/github/minecraftcursedlegacy/testmixin/MixinPlayer.java:8]: it's in the src/test set

Comment (valoeghese) [reply src/test/java/io/github/minecraftcursedlegacy/testmixin/MixinPlayer.java:8]: It doesn't do anything but it saves having to add a MixinPlayer class next time I want to make a test mixin related to player actions

File: src/main/java/io/github/minecraftcursedlegacy/mixin/MixinItemEntity.java (hunk 9 above)
Comment (valoeghese) [reply src/main/java/io/github/minecraftcursedlegacy/mixin/MixinItemEntity.java]: yes

**Comment by valoeghese** [comment]
Should be fine now

**Comment by valoeghese** [comment]
wait lemme do juuz's doc change

**Comment by valoeghese** [comment]
ok done

**Review by williambl — APPROVED** [review]

**Pull request merged**